app/src/main/java/com/github/sdpteam15/polyevents/model/entity/Zone.kt

```diff
@@ -175,15 +176,23 @@ data class Zone(
     }
 
     private fun edgeIsInZone(edges: RouteEdge): Boolean {
-        val mid = divide(plus(edges.start!!.toLatLng(), edges.end!!.toLatLng()), 2.0)
+        val list = mutableListOf(
+            edges.start!!.toLatLng(),
+            edges.end!!.toLatLng(),
+        )
+        if(edges.start!!.areaId != zoneId)
+            list.remove(edges.end!!.toLatLng())
+        if(edges.end!!.areaId != zoneId)
+            list.remove(edges.start!!.toLatLng())
+        val mean = mean(list)
         for (l in getZoneCoordinates()) {
-            val AM = minus(mid, l[0])
+            val AM = minus(mean, l[0])
             val AB = minus(l[1], l[0])
-            val BM = minus(mid, l[1])
+            val BM = minus(mean, l[1])
             val BC = minus(l[2], l[1])
             //https://math.stackexchange.com/questions/190111/how-to-check-if-a-point-is-inside-a-rectangle
```

Comment:
Maybe put this the link into the method documentation Robin, and add a doc stating briefly what the method does.